firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigMetadataClient.java

```diff
@@ -249,6 +266,72 @@ void setBackoffMetadata(int numFailedFetches, Date backoffEndTime) {
     }
   }
 
+  public void setCustomSignals(Map<String, String> newCustomSignals) {
+    synchronized (customSignalsLock) {
+      // Retrieve existing custom signals
+      Map<String, String> existingCustomSignals = getCustomSignals();
+
+      for (Map.Entry<String, String> entry : newCustomSignals.entrySet()) {
+        String key = entry.getKey();
+        String value = entry.getValue();
+
+        // Validate key and value length
+        if (key.length() > CUSTOM_SIGNALS_MAX_KEY_LENGTH
+            || (value != null && value.length() > CUSTOM_SIGNALS_MAX_STRING_VALUE_LENGTH)) {
+          Log.w(
+              TAG,
+              String.format(
+                  "Invalid custom signal: Custom signal keys must be %d characters or less, and values must be %d characters or less.",
+                  CUSTOM_SIGNALS_MAX_KEY_LENGTH, CUSTOM_SIGNALS_MAX_STRING_VALUE_LENGTH));
+          return;
+        }
+
+        // Merge new signals with existing ones, overwriting existing keys.
+        // Also, remove entries where the new value is null.
+        if (value != null) {
+          existingCustomSignals.put(key, value);
+        } else {
+          existingCustomSignals.remove(key);
+        }
+      }
+
+      // Check if the map has actually changed and the size limit
+      if (existingCustomSignals.equals(getCustomSignals())) {
+        return;
+      }
+      if (existingCustomSignals.size() > CUSTOM_SIGNALS_MAX_COUNT) {
+        Log.w(
+            TAG,
+            String.format(
+                "Invalid custom signal: Too many custom signals provided. The maximum allowed is %d.",
+                CUSTOM_SIGNALS_MAX_COUNT));
+        return;
+      }
+
+      frcMetadata
+          .edit()
+          .putString(CUSTOM_SIGNALS, new JSONObject(existingCustomSignals).toString())
```

Comment:
consider adding some debug logs `Log.d` to indicate what the final set of custom signals is, especially since there's no `get` API right now

Reply:
Added debug logs for the final custom signals after updates and during fetches.